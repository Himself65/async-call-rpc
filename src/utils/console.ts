//#region Console
/**
 * The minimal Console interface that AsyncCall needs.
 * @public
 */
export interface Console {
    debug?(...args: unknown[]): void
    log(...args: unknown[]): void
    groupCollapsed?(...args: unknown[]): void
    groupEnd?(...args: unknown[]): void
    error?(...args: unknown[]): void
}
