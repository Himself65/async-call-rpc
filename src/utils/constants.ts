export const isString = (x: unknown): x is string => typeof x === 'string'
export const isBoolean = (x: unknown): x is boolean => typeof x === 'boolean'
export const isFunction = (x: unknown): x is Function => typeof x === 'function'
export const isObject = (params: any): params is object => typeof params === 'object' && params !== null
export const ERROR = 'Error'
export const undefined = void 0
export const Object_setPrototypeOf = Object.setPrototypeOf
export const Promise_resolve = <T>(x: T) => Promise.resolve(x)
export const isArray = Array.isArray
export const replayFunction = () => '() => replay()'
