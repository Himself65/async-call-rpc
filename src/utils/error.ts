class CustomError extends Error {
    constructor(public name: string, message: string, public code: number, public stack: string) {
        super(message)
    }
}
/** These Error is defined in ECMAScript spec */
const errors: Record<string, typeof EvalError> = {
    Error,
    EvalError,
    RangeError,
    ReferenceError,
    SyntaxError,
    TypeError,
    URIError,
}
export const DOMExceptionHeader = 'DOMException:'
/**
 * AsyncCall support somehow transfer ECMAScript Error
 */
export function RecoverError(type: string, message: string, code: number, stack: string): Error {
    try {
        if (type.startsWith(DOMExceptionHeader) && DOMException) {
            const [, name] = type.split(DOMExceptionHeader)
            return new DOMException(message, name)
        } else if (type in errors) {
            const e = new errors[type](message)
            e.stack = stack
            Object.assign(e, { code })
            return e
        } else {
            return new CustomError(type, message, code, stack)
        }
    } catch {
        return new Error(`E${code} ${type}: ${message}\n${stack}`)
    }
}
export function removeStackHeader(stack = '') {
    return stack.replace(/^.+\n.+\n/, '')
}
function getDOMException(): { new (message: string, name: string): any } | undefined {
    const x = Reflect.get(globalThis, 'DOMException')
    if (typeof x === 'function') return x
    return undefined
}
export const DOMException = getDOMException()
