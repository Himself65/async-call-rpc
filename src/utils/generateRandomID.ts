export const generateRandomID = () => Math.random().toString(36).slice(2)
