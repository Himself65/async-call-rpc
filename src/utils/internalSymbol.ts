export const AsyncCallIgnoreResponse = Symbol.for('AsyncCall: This response should be ignored.')
