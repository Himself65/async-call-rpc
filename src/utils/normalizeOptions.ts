import { AsyncCallOptions, AsyncCallStrictJSONRPC } from '../Async-Call'
import { isBoolean } from './constants'
export const normalizeLogOptions = (log: NonNullable<AsyncCallOptions['log']>) => {
    if (!isBoolean(log)) {
        const { beCalled, localError, remoteError, type, requestReplay, sendLocalStack } = log
        return [beCalled, localError, remoteError, type !== 'basic', requestReplay, sendLocalStack] as const
    }
    if (log) return [true, true, true, true] as const
    return [] as const
}

export const normalizeStrictOptions = (strict: NonNullable<AsyncCallOptions['strict']>) => {
    if (!isBoolean(strict)) {
        const { methodNotFound, unknownMessage } = strict
        return [methodNotFound, unknownMessage] as const
    }
    return [strict, strict] as const
}
