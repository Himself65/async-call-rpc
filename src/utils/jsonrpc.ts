import { DOMException, DOMExceptionHeader } from './error'
import { ErrorMapFunction } from '../Async-Call'

export const jsonrpc = '2.0'
export type ID = string | number | null | undefined
/**
 * JSONRPC Request object.
 */
export interface Request
    extends Readonly<{
        jsonrpc: typeof jsonrpc
        id?: ID
        method: string
        params: readonly unknown[] | object
        remoteStack?: string
    }> {}

export function Request(id: ID, method: string, params: readonly unknown[] | object, remoteStack?: string): Request {
    const x: Request = { jsonrpc, id, method, params, remoteStack }
    deleteUndefined(x, 'id')
    deleteFalsy(x, 'remoteStack')
    return x
}

/**
 * JSONRPC SuccessResponse object.
 */
export interface SuccessResponse
    extends Readonly<{
        jsonrpc: typeof jsonrpc
        id?: ID
        result: unknown
    }> {}
export function SuccessResponse(id: ID, result: unknown): SuccessResponse {
    const x: SuccessResponse = { jsonrpc, id, result }
    deleteUndefined(x, 'id')
    return x
}

/**
 * JSONRPC ErrorResponse object.
 * @public
 */
export interface ErrorResponse<E = unknown>
    extends Readonly<{
        jsonrpc: typeof jsonrpc
        id?: ID
        error: Readonly<{ code: number; message: string; data?: E }>
    }> {}

export function ErrorResponse<T>(id: ID, code: number, message: string, data?: T): ErrorResponse<T> {
    if (id === undefined) id = null
    code = Math.floor(code)
    if (Number.isNaN(code)) code = -1
    const x: ErrorResponse<T> = { jsonrpc, id, error: { code, message, data } }
    deleteUndefined(x.error, 'data')
    return x
}
// Pre defined error in section 5.1
ErrorResponse.ParseError = <T>(e: unknown, mapper: ErrorMapFunction<T>): ErrorResponse<T> => {
    const obj = ErrorResponseMapped({} as any, e, mapper)
    const o = obj.error as Mutable<ErrorResponse['error']>
    o.code = -32700
    o.message = 'Parse error'
    return obj
}

// Not using.
// InvalidParams -32602 'Invalid params'
// InternalError -32603 'Internal error'
ErrorResponse.InvalidRequest = (id: ID) => ErrorResponse(id, -32600, 'Invalid Request')
ErrorResponse.MethodNotFound = (id: ID) => ErrorResponse(id, -32601, 'Method not found')

type AsyncCallErrorDetail = {
    stack?: string
    type?: string
}
export function ErrorResponseMapped<T>(request: Request, e: unknown, mapper: ErrorMapFunction<T>): ErrorResponse<T> {
    const { id } = request
    const { code, message, data } = mapper(e, request)
    return ErrorResponse(id, code, message, data)
}

export const defaultErrorMapper = (stack = '', code = -1): ErrorMapFunction<AsyncCallErrorDetail> => (e) => {
    let message = ''
    if (isObject(e) && hasKey(e, 'message') && typeof e.message === 'string') message = e.message
    let type = toString('Error', () => (e as any)?.constructor?.name)
    if (DOMException && e instanceof DOMException) type = DOMExceptionHeader + e.name
    if (typeof e === 'string' || typeof e === 'number' || typeof e === 'boolean' || typeof e === 'bigint') {
        type = 'Error'
        message = String(e)
    }
    const data: AsyncCallErrorDetail = stack ? { stack, type } : { type }
    return { code, message, data }
}

/**
 * A JSONRPC response object
 */
export type Response = SuccessResponse | ErrorResponse

export function isJSONRPCObject(data: any): data is Response | Request {
    if (!isObject(data)) return false
    if (!hasKey(data, 'jsonrpc')) return false
    if (data.jsonrpc !== jsonrpc) return false
    if (hasKey(data, 'params')) {
        const params = (data as Request).params
        if (!Array.isArray(params) && !isObject(params)) return false
    }
    return true
}

export function isObject(params: any): params is object {
    return typeof params === 'object' && params !== null
}

export function hasKey<T, Q extends string>(
    obj: T,
    key: Q,
): obj is T &
    {
        [key in Q]: unknown
    } {
    return key in obj
}

function toString(def: string, val: () => any) {
    let str = def
    try {
        str = val()
    } catch {}
    if (typeof str !== 'string') return def
    return str
}
function deleteUndefined<O>(x: O, key: keyof O) {
    if (x[key] === undefined) delete x[key]
}
function deleteFalsy<T>(x: T, key: keyof T) {
    if (!x[key]) delete x[key]
}
type Mutable<T> = { -readonly [key in keyof T]: T[key] }
