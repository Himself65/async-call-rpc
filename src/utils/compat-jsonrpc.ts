import { ErrorResponse, Request, SuccessResponse } from './jsonrpc'

export type ID = string | number | null | undefined
type CompatRequest = [type: 0, method: string, params: readonly unknown[] | object, id?: ID, stack?: string]
// [normal: 0, id, method, ...params]
// [+stack: 1, id, method, stack, ...params]
// [notify: 2, method, ...params]
// [+stack: 3, stack, ...params]
type CompatSuccessResponse = [type: 1, id: ID, result?: unknown]
type CompatErrorResponse = [type: 2, id: ID, code: number, message: string, data?: unknown]
export const CompressRequest = (req: Request): CompatRequest => {
    const x: CompatRequest = [0, req.method, req.params]
    if (req.id !== undefined) x[3] = req.id
    if (req.remoteStack) x[4] = req.remoteStack
    return x
}
export const DecompressRequest = ([, methods, param, id, stack]: CompatRequest) => Request(id, methods, param, stack)
export const CompressSuccessResponse = ({ result, id }: SuccessResponse): CompatSuccessResponse => {
    const x: CompatSuccessResponse = [1, id]
    if (result !== undefined) x[2] = result
    return x
}
export const DecompressSuccessResponse = ([, ...r]: CompatSuccessResponse) => SuccessResponse(...r)
export const CompressErrorResponse = ({ error: { code, message, data }, id }: ErrorResponse): CompatErrorResponse => {
    const x: CompatErrorResponse = [2, id, code, message]
    if (data !== undefined) x[4] = data
    return x
}
export const DecompressErrorResponse = ([, ...y]: CompatErrorResponse): ErrorResponse => ErrorResponse(...y)
export const Decompress = (x: CompatRequest | CompatErrorResponse | CompatSuccessResponse) => {
    if (x[0] == 0) return DecompressRequest(x)
    if (x[0] == 1) return DecompressSuccessResponse(x)
    return DecompressErrorResponse(x)
}
