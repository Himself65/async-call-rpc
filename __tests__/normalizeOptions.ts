import { normalizeLogOptions, normalizeStrictOptions } from '../src/utils/normalizeOptions'

test('normalize log options', () => {
    expect(normalizeLogOptions(true)).toMatchSnapshot()
    expect(normalizeLogOptions(false)).toMatchSnapshot()
    expect(normalizeLogOptions({ beCalled: true, localError: false })).toMatchSnapshot()
})

test('normalize strict options', () => {
    expect(normalizeStrictOptions(true)).toMatchSnapshot()
    expect(normalizeStrictOptions(false)).toMatchSnapshot()
    expect(normalizeStrictOptions({ noUndefined: true, methodNotFound: false })).toMatchSnapshot()
})
