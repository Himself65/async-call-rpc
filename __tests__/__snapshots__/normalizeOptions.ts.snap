// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize log options 1`] = `
Array [
  true,
  true,
  true,
  true,
  true,
  true,
]
`;

exports[`normalize log options 2`] = `
Array [
  true,
  true,
  true,
  true,
]
`;

exports[`normalize log options 3`] = `Array []`;

exports[`normalize log options 4`] = `
Array [
  true,
  false,
  undefined,
  true,
  undefined,
  undefined,
]
`;

exports[`normalize strict options 1`] = `
Array [
  true,
  true,
]
`;

exports[`normalize strict options 2`] = `
Array [
  false,
  false,
]
`;

exports[`normalize strict options 3`] = `
Array [
  false,
  false,
]
`;

exports[`normalize strict options 4`] = `
Array [
  false,
  true,
]
`;

exports[`normalize strict options 5`] = `
Array [
  true,
  false,
]
`;

exports[`normalize strict options 6`] = `
Array [
  true,
  true,
]
`;

exports[`normalize strict options 7`] = `
Array [
  false,
  undefined,
]
`;

exports[`normalize strict options 8`] = `
Array [
  true,
  undefined,
]
`;

exports[`normalize strict options 9`] = `
Array [
  undefined,
  false,
]
`;

exports[`normalize strict options 10`] = `
Array [
  undefined,
  true,
]
`;
